src/server/search/aggregator.cc

```diff
@@ -80,22 +80,41 @@ void Aggregator::DoSort(std::string_view field, bool descending) {
       desc -> false
   */
   auto comparator = [&](const DocValues& l, const DocValues& r) {
-    auto l_it = l.find(field);
-    auto r_it = r.find(field);
-
-    // If some of the values is not present
-    if (l_it == l.end() || r_it == r.end()) {
-      return l_it != l.end();
+    for (const auto& [field, order] : sort_params.fields) {
+      auto l_it = l.find(field);
+      auto r_it = r.find(field);
+
+      // If some of the values is not present
+      if (l_it == l.end() || r_it == r.end()) {
+        if (l_it == l.end() && r_it == r.end()) {
+          continue;
+        }
+        return l_it != l.end();
+      }
+
+      const auto& lv = l_it->second;
+      const auto& rv = r_it->second;
+      if (lv == rv) {
+        continue;
+      }
+      return order == SortParams::SortOrder::ASC ? lv < rv : lv > rv;
     }
-
-    auto& lv = l_it->second;
-    auto& rv = r_it->second;
-    return !descending ? lv < rv : lv > rv;
+    return false;
   };
 
-  std::sort(result.values.begin(), result.values.end(), std::move(comparator));
+  auto& values = result.values;
+  if (sort_params.SortAll()) {
+    std::sort(values.begin(), values.end(), comparator);
+  } else {
+    DCHECK_GE(sort_params.max, 0);
+    const size_t limit = std::min(values.size(), size_t(sort_params.max));
+    std::partial_sort(values.begin(), values.begin() + limit, values.end(), comparator);
+    values.resize(limit);
+  }
 
-  result.fields_to_print.insert(field);
+  for (auto& field : sort_params.fields) {
+    result.fields_to_print.insert(field.first);
```

Comment:
I believe instead of a loop you should use insert(begin, end)

Reply:
The problem is that` sort_params.fields` is vector of `pair<string, SortOrder>`, `result.fields_to_print` is set of `string_view`. So we need to split `sort_params` to two vectors. I don't think it worth to do it so

Reply:
you are right